### ISSUE TITLE
ADAL 4.3.0 with ADFS 3.0: The server returned an invalid response

### ISSUE THREAD

**Issue by rodro75** [AzureAD/azure-activedirectory-library-for-dotnet#1354]
**Which Version of ADAL are you using ?**
ADAL 4.3.0

**Which platform has the issue?**
net45

**What authentication flow has the issue?**
* Desktop / Mobile
    * Interactive
    * Integrated Windows Auth

After upgrading from 3.19.8 my WPF client stopped working.
The call to `AuthenticationContext.AcquireTokenAsync` now throws an `AdalServiceException` after successful login, saying that it received an invalid response from the ADFS server.

Looking at what's going on with Fiddler I can see that the new version sends an extra queryString parameter to the ADFS server: `response_mode=form_post`, which wasn't sent before.
Everything else looks pretty much the same as before, but the library fails right before making the POST to `/adfs/oauth2/token`.

**Comment by rodro75** [comment]
Here are the HTTP dumps of both a working flow with 3.19.8 and a non working one with 4.3.0.

[Traffic.zip](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/files/2515147/Traffic.zip)

**Comment by jmprieur** [comment]
@henrik-me @bgavrilMS @MarkZuber FYI

**Comment by jmprieur** [comment]
Thanks for the heads-up, @rodro75

**Comment by MarkZuber** [comment]
Hi @rodro75, do you happen to have debug logs from your execution that I could get from you?  That would be rather helpful in my ability to rapidly diagnose your issue.  thanks!

**Comment by henrik-me** [comment]
@reed1995 : fyi
@rodro75 : which version of ADFS are you using?

**Comment by rodro75** [comment]
@henrik-me ADFS 3.0 on Windows Server 2012 R2

@MarkZuber Here is the debug log:

```
Information 2018-10-30T11:19:27.8669740Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: ADAL PCL.Desktop with assembly version '4.3.0.0', file version '4.3.0.0' and informational version '4.3.0' is running...
Information 2018-10-30T11:19:27.8699852Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: === Token Acquisition started:
	Authority: https://sts.dvr.local/adfs/
	Resource: https://milano.dvritalia.it:8001/adfs-app1
	ClientId: D38F0FA2-3E85-4862-AF80-FEB9D00BE1A0
	CacheType: null
	Authentication Target: User
	
Verbose 2018-10-30T11:19:27.8759816Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Loading from cache.
Verbose 2018-10-30T11:19:27.8830702Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Looking up cache for a token...
Information 2018-10-30T11:19:27.8860566Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: No matching token was found in the cache
Information 2018-10-30T11:19:27.8870658Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Checking MSAL cache for user token cache
Verbose 2018-10-30T11:19:27.8880562Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Looking up cache for a token...
Information 2018-10-30T11:19:27.8890551Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: No matching token was found in the cache
Information 2018-10-30T11:19:27.8890551Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Checking MSAL cache for user token cache
Verbose 2018-10-30T11:19:44.8644582Z: 00000000-0000-0000-0000-000000000000 - AdalLoggerBase.cs:  WebBrowser state: IsBusy: True, ReadyState: Complete, Created: True, Disposing: False, IsDisposed: False, IsOffline: False
Verbose 2018-10-30T11:19:44.8664430Z: 00000000-0000-0000-0000-000000000000 - AdalLoggerBase.cs:  WebBrowser state (after Stop): IsBusy: False, ReadyState: Complete, Created: True, Disposing: False, IsDisposed: False, IsOffline: False
Error 2018-10-30T11:19:44.9670134Z: 02a83395-c5f7-423f-8725-7237363b5380 - AdalLoggerBase.cs: Microsoft.IdentityModel.Clients.ActiveDirectory.AdalServiceException: The authorization server returned an invalid response
   in Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenInteractiveHandler.VerifyAuthorizationResult()
   in Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenInteractiveHandler.<PreTokenRequestAsync>d__10.MoveNext()
--- Fine traccia dello stack da posizione precedente dove è stata generata l'eccezione ---
   in System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   in System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   in System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   in Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.<RunAsync>d__53.MoveNext()
	ErrorCode: authentication_failed
	StatusCode: 0
```

For completeness here is my dead simple code:

```
            var authority = "https://sts.dvr.local/adfs";
            var resourceURI = "https://milano.dvritalia.it:8001/adfs-app1";
            var clientID = "D38F0FA2-3E85-4862-AF80-FEB9D00BE1A0";
            var redirectURI = "https://milano.dvritalia.it:8001/adfs-app1/ClientReturn.aspx";

            var authenticationContext = new AuthenticationContext(authority, false);
            var result = await authenticationContext.AcquireTokenAsync(resourceURI, clientID,
                new Uri(redirectURI),
                new PlatformParameters(PromptBehavior.Auto));
```

The ADFS 3.0 server is configured to only allow Forms Authentication.

**Comment by trwalke** [comment]
@rodro75 So far I have been unable to reproduce this error on our side, however am going to keep investigating. Is there a reason that you need to have only the Forms Authentication turned on for the ADFS 3.0 server?

**Comment by trwalke** [comment]
@rodro75 Unfortunatly, I have been unable to reproduce the behavior you are seeing here. I have tried setting our ADFS v 3.0 to only allow Forms Authentication and still was not able to reproduce. Perhaps there is a configuration in your ADFS server that does not work with the Form Post Response Mode that we have added. 
Please visit https://aka.ms/msidgethelp for more assistance. 

If you have more information to help us investigate please post it here.
Thank You For Your Patience

**Comment by kennethahn** [comment]
I am adding ADFS support to a product at work, and started out with a Hello ADFS app like the example provided here. Also using 2012R2 and ADFS 3.0 - with latest ADAL libs. After spending a day not getting it to work, I downgraded library to 3.19.8 and voilá - everything works.

**Comment by jmprieur** [comment]
Reopening (See Kennethahn's report)
cc: @trwalke @MarkZuber 

@kennethahn : which Hello ADFS app is it? can you please send us a link?

**Comment by kennethahn** [comment]
@jmprieur It is the same 5 lines of code that is in all howto's. :)
I have pasted it here: https://pastebin.com/Y1TfQmqD

**Comment by trwalke** [comment]
@kennethahn Are there any proxies that you are using that may interfere with the response_mode=form_post? Also, is the server 2012 updated to the latest version? ADFS v3?

So far, we are unable to reproduce your error with the new response_mode=form_post, forms only authentication, the latest version of ADAL and 

![adfsv3](https://user-images.githubusercontent.com/30090357/49118497-4d700c00-f259-11e8-8d3d-bd464d6d1c86.png)
![server2012](https://user-images.githubusercontent.com/30090357/49118498-4d700c00-f259-11e8-9918-5e90cd31410b.png)

**Comment by billvolt** [comment]
I'm also targeting ADFS 3.0 on Windows Server 2012 R2.  I found old Windows Forms or WPF code samples authenticated properly while newer samples did not.  In my testing Microsoft.IdentityModel.Clients.ActiveDirectory versions as late as 3.19.8 work consistently while later versions (including 4.5.1) fail consistently.

As did rodro75, I noted the response_mode=form_post in the querystring captured by Fiddler.  I also noted that only the newer build attempts an HTTP GET against the redirect URL provided by ADFS in the Location header of the fourth response from /adfs/oauth2/authorize.  Subsequently, the newer build fails to access my Web API.

Running Microsoft.IdentityModel.Clients.ActiveDirectory 4.4.2 against ADFS 4.0, Fiddler shows only three calls to /adfs/oauth2/authorize.  The third one has a body that contains an HTML form set up to post to the redirect URL.  The redirect URL is not contained in a header.  ADFS 3.0 does not return bodies in responses from /adfs/oauth2/authorize after the first.

Below are partial captures of the fourth call to /adfs/oauth2/authorize and subsequent attempt to follow the URL from the Location header.

`GET https://fs.adfslab3.local/adfs/oauth2/authorize?resource=https%3A%2F%2Fabc.xyz.net%2FMyWebAPIsample%2F&client_id=09c9a8a2-6bf1-427d-89ba-45c2c02bb9fc&response_type=code&haschrome=1&redirect_uri=http%3A%2F%2Fanarbitraryreturnuri%2F&response_mode=form_post&client-request-id=fd6e51a0-f5ad-4481-9045-c981b02dc6f5&prompt=attempt_none&x-client-SKU=PCL.Desktop&x-client-Ver=4.5.1.0&x-client-CPU=x64&x-client-OS=Microsoft+Windows+NT+6.2.9200.0 HTTP/1.1`

```
HTTP/1.1 302 Found
Cache-Control: no-cache,no-store
Pragma: no-cache
Content-Length: 0
Content-Type: text/html; charset=utf-8
Expires: -1
Location: http://anarbitraryreturnuri:80/?code=rmGjA-e270-0Q0KeUrY6FQ.3KIo5jfE1ggaAKkR_vSk8vP1h-U.EP9fOjHz0BqM_RJUXiJJrvgMZ7g6FqdSGH5yPdTzM8FRfl1X0hM920m8YVEUWG1QMBG7Iwfx9BymcYd2q05ykvO-EPoaAao4Mw9oW2EDB-ozy3-PyQHcCVd_9nv7smpjj2IyCP8roVRFs_xpKeEWzWCl-TbZp_K9jgIVZHmp_RcQq6w1vrTRmfkvMeWLYeDKQv46auqJmdYWIdcnYVJJ00CFChqxNSE7GttGdZ4Egn6fxbocFtRkguGj3z4s0Q5DyIAOrX-_lnysIN-eNQEp7Ld416i5t3JRDbIbkBtLuoxFe0Q8X5c3RXsnjZJsKmx8TLKjRrv8wNlTANH8XHFD-A&client-request-id=fd6e51a0-f5ad-4481-9045-c981b02dc6f5
Server: Microsoft-HTTPAPI/2.0
P3P: CP="ADFS doesn't have P3P policy, please contact your site's admin for more details."
Set-Cookie: MSISAuthenticated=NC8xOC8yMDE5IDc6NTY6MDggUE0=; path=/adfs; HttpOnly; Secure
Set-Cookie: MSISLoopDetectionCookie=MjAxOS0wNC0xODoxOTo1NjowNFpcMg==; path=/adfs; HttpOnly; Secure
Date: Thu, 18 Apr 2019 19:56:08 GMT

```


`GET http://anarbitraryreturnuri/?code=rmGjA-e270-0Q0KeUrY6FQ.3KIo5jfE1ggaAKkR_vSk8vP1h-U.EP9fOjHz0BqM_RJUXiJJrvgMZ7g6FqdSGH5yPdTzM8FRfl1X0hM920m8YVEUWG1QMBG7Iwfx9BymcYd2q05ykvO-EPoaAao4Mw9oW2EDB-ozy3-PyQHcCVd_9nv7smpjj2IyCP8roVRFs_xpKeEWzWCl-TbZp_K9jgIVZHmp_RcQq6w1vrTRmfkvMeWLYeDKQv46auqJmdYWIdcnYVJJ00CFChqxNSE7GttGdZ4Egn6fxbocFtRkguGj3z4s0Q5DyIAOrX-_lnysIN-eNQEp7Ld416i5t3JRDbIbkBtLuoxFe0Q8X5c3RXsnjZJsKmx8TLKjRrv8wNlTANH8XHFD-A&client-request-id=fd6e51a0-f5ad-4481-9045-c981b02dc6f5 HTTP/1.1`

```
HTTP/1.1 502 Fiddler - DNS Lookup Failed
Date: Thu, 18 Apr 2019 19:56:09 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 15:56:09.539
```

Perhaps this ADAL 4.x behavior of following the redirect URL provides a useful clue?

**Comment by billvolt** [comment]
I can provide a sample app and Windows Server 2012 R2 virtual machine for analysis.

**Comment by trwalke** [comment]
@billvolt that would be perfect. you can send me an email @ trwalke@microsoft.com with all of the details of the repro. In the mean time, I will poke around on my side to see if I can diagnose the issue again. Thanks for all of the additional info.

**Comment by bgavrilMS** [comment]
This is almost 1 year old, can we close this ?

**Comment by bgavrilMS** [comment]
Closing, please reopen if there are more details.

**Comment by StevenDevooght** [comment]
Hi,

We were also facing the same issue using ADFS 3.0 on Windows server 2012 R2. After authenticating on the ADFS page we are redirected to our callback_uri with an authorization_code appended in the querystring.

The problem is the 'code' querystring part gets lost on [this line](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/68d7dea3643e075be85abbca1ab88ba614465541/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/net45/WindowsFormsWebAuthenticationDialogBase.cs#L296).

**Comment by billvolt** [comment]
I've identified a fix and submitted pull request #1741.
